### ISSUE TITLE
Nested parallelism with xgboost

### ISSUE THREAD

**Issue by adatum** [topepo/caret#924]
Should `caret` and/or the parallel backend avoid nested parallelism?

As described in the [documentation](http://topepo.github.io/caret/parallel-processing.html), using certain functions in combination with using a parallel backend with P processors results in P^2 processes. I found out that xgboost (xgbLinear, xgbTree, etc) happens to be one of these functions.

This results in a P-core processor being overloaded by a factor of P, greatly increasing computation time (5x slowdown in one of my tests).

Workarounds:

- have xgboost use one thread (via `nthread=1` parameter in `caret::train`, or `tuneList=list(xgbLinear=caretModelSpec(method="xgbLinear", nthread=1))` in `caretEnsemble::caretList`) while using a parallel backend (`doMC`, `doFuture`, etc) and `allowParallel=TRUE`
- let xgboost parallelize which it does by default, and set caret's `allowParallel=FALSE` and/or don't use a parallel backend

It seems there may be some performance difference between the two options above, as indicated [here](https://stackoverflow.com/questions/39528392/parallel-processing-with-xgboost-and-caret).

I tried using the `doFuture` backend since @HenrikBengtsson mentioned in [another issue](https://github.com/topepo/caret/issues/449) that it would automatically prevent this problem, but at least with xgboost I still counted P^2 processes in `htop`.

**Comment by topepo** [comment]
> Should caret and/or the parallel backend avoid nested parallelism?

Yes. From the [documentation](https://topepo.github.io/caret/parallel-processing.html): 

> Also, for `rfe` and  `sbf`, these functions may call  `train` for some models. In this case, registering *M* workers will actually invoke *M*^2^ total processes.

That's why most of `caret`'s functions have an `allowParallel` option.

This is also true for models like `ranger` or `xgboost` that have their own parallelism. 

> It seems there may be some performance difference between the two options above, as indicated here.

My response in that link sums up my feelings on the matter. 

A more modern benchmark [backs up](http://appliedpredictivemodeling.com/blog/2018/1/17/parallel-processing) those ideas; if you are resampling, you are much better off with no parallelism in the individual models and running the resamples in parallel. If it is a single model fit with no resampling or tuning then use the model's functionality. 

During resampling or tuning, the individual model has to start and stop its parallel infrastructure a ton of times (since it is recalled repeatedly). You parallelize  to longest running part of the process.

**Comment by adatum** [comment]
Thanks for the link with the benchmarking; it helps understand much better the different approaches to parallelization and their performance impact.

The issue is that the combined defaults of both `xgboost` and `caret` result in nested parallelism. Perhaps `caret` can set either the `nthread` or `allowParallel` default parameter to serial processing when using the `xgboost` model to avoid this? 

Of course the user could still set their own parameters as now, but at least the defaults would be sensible.

I suspect the typical use case with `caret` would be resampling, and thus benefit more from setting `nthread=1`?

---

Aside, if training several models with `caretList`, `tuneList` can be used to set individual model parameters like `nthread`. But I don't know if `allowParallel` can be set per model, since it is defined by `trainControl` which is given for the whole `caretList`.

**Comment by topepo** [comment]
It bugs me that the default is to use all cores (or at least > 1). I'm surprised that CRAN allows that. 

It is possible to deal with this in the `caret` model code but I'd like to avoid it. It would involve intercepting the `...` finding the relevant  argument, change it if a parallel backend is registered, setup an argument list, merging `...` back in, and change the invokation to use `do.call`. That's ripe for errors and would probably cause more issues than it solves.
